src/coreclr/src/System.Private.CoreLib/src/System/RuntimeHandles.cs

```diff
@@ -1273,27 +1273,50 @@ public bool Equals(ModuleHandle handle)
         private static void ValidateModulePointer(RuntimeModule module)
         {
             // Make sure we have a valid Module to resolve against.
-            if (module == null)
-                throw new InvalidOperationException(SR.InvalidOperation_NullModuleHandle);
+            if (module is null)
+            {
+                // Local function to allow inlining of simple null check.
+                ThrowInvalidOperationException();
+            }
+
+            [StackTraceHidden]
+            [DoesNotReturn]
+            static void ThrowInvalidOperationException() => throw new InvalidOperationException(SR.InvalidOperation_NullModuleHandle);
         }
 
         // SQL-CLR LKG9 Compiler dependency
         public RuntimeTypeHandle GetRuntimeTypeHandleFromMetadataToken(int typeToken) { return ResolveTypeHandle(typeToken); }
         public RuntimeTypeHandle ResolveTypeHandle(int typeToken)
         {
-            return new RuntimeTypeHandle(ResolveTypeHandleInternal(GetRuntimeModule(), typeToken, null, null));
+            return new RuntimeTypeHandle(ResolveTypeHandleInternal(GetRuntimeModule(), typeToken));
         }
         public RuntimeTypeHandle ResolveTypeHandle(int typeToken, RuntimeTypeHandle[]? typeInstantiationContext, RuntimeTypeHandle[]? methodInstantiationContext)
         {
             return new RuntimeTypeHandle(ModuleHandle.ResolveTypeHandleInternal(GetRuntimeModule(), typeToken, typeInstantiationContext, methodInstantiationContext));
         }
 
+        internal static RuntimeType ResolveTypeHandleInternal(RuntimeModule module, int typeToken)
+        {
+            try
+            {
+                RuntimeType? type = null;
+                ResolveType(new QCallModule(ref module), typeToken, null, 0, null, 0, ObjectHandleOnStack.Create(ref type));
+                return type!;
+            }
+            catch (Exception)
+            {
+                ValidateModulePointer(module);
+                if (!ModuleHandle.GetMetadataImport(module).IsValidToken(typeToken))
+                    throw new ArgumentOutOfRangeException(nameof(typeToken),
+                        SR.Format(SR.Argument_InvalidToken, typeToken, new ModuleHandle(module)));
+                throw;
+            }
+        }
+
         internal static RuntimeType ResolveTypeHandleInternal(RuntimeModule module, int typeToken, RuntimeTypeHandle[]? typeInstantiationContext, RuntimeTypeHandle[]? methodInstantiationContext)
         {
-            ValidateModulePointer(module);
-            if (!ModuleHandle.GetMetadataImport(module).IsValidToken(typeToken))
-                throw new ArgumentOutOfRangeException(nameof(typeToken),
-                    SR.Format(SR.Argument_InvalidToken, typeToken, new ModuleHandle(module)));
+            Debug.Assert((typeInstantiationContext is not null && typeInstantiationContext.Length > 0) ||
```

Comment:
This assert does not hold when this is called via the public `ResolveTypeHandle(Int32, RuntimeTypeHandle[], RuntimeTypeHandle[])` method.

Reply:
I am wondering why we need to have both public and internal versions of these methods. It is trivial to hop from RuntimeTypeHandle to RuntimeType and back, and same for Module. I think it would be nice to delete the `Internal` layer.

Reply:
Unified